Remove use of `pkg/pools` in archive

dmcgowan opened the pull request:
Removes unnecessary use of pooling and prefers use of buffer pool on copy rather than pooling bufio. In some cases the bufio was used and the pool copy was still used which unnecessarily got a buffer which would not be used.

dmcgowan committed:
Remove use of pools in archive

Signed-off-by: Derek McGowan <derek@mcg.dev>

dmcgowan committed:
Fix chrootarchive test

After the untar errors, the reader must complete in order to fill
the buffer used by the subsequent check.

Signed-off-by: Derek McGowan <derek@mcg.dev>

dmcgowan committed:
Add pool for archive decompress stream

Cleanup decompress logic and add a pool. The close logic should be
custom defined for each compression type since they have different
close interfaces.

Signed-off-by: Derek McGowan <derek@mcg.dev>

dmcgowan commented:
Update the DecompressStream logic to match what is done in containerd, it has the buffer pool now and cleans up a case where zstd decompression was not getting `Close` called.

dmcgowan commented:
The previous test failures were both related to the tests erroneously relying on the bufio to fill up the read buffer while now the pipe reader may return smaller chunks. Switching to `io.ReadFull` when a certain number of bytes is needed is the correct way to handle that. The chrootarchive failure was caused by abandoning a tee reader then assuming the write buffer was filled, so the correct thing to do is drain the reader before using the write buffer. There seem to be a few more cases like that I need to look at but I will mark this as ready for review.

vvoland commented on pkg/archive/archive.go:
```diff
@@ -231,13 +231,51 @@ func (r *readCloserWrapper) Close() error {
+		return 0, io.EOF
+	}
+	n, err = r.buf.Read(p)
+	if err == io.EOF {
```
Won't this leak the buffer if read fails with other errors? Couldn't this just be `err != nil`?

cpuguy83 replied:
Guess this should have a closer to return the buffer

dmcgowan replied:
In that case it will just get GC'ed. It is pointless to return to the pool on other errors since the higher level operation will end. Am I missing something else which will hold other a reference?

This implementation is intentionally the same as containerd's. We can probably unify the two, the only difference is we don't (intentionally) support xz.

thaJeztah commented:
Silly question; the existing pkg/pools package had some test-coverage on those pools (although I had a quick peek at some of those, and it may be very limited). With some of these being moved internal it looks like we don't have separate coverage for their replacements, other than existing tests covering them as part of where they're used.

Will we be reducing coverage with this, or are we confident that the existing tests were already covering the functionality enough?

dmcgowan commented:
> Will we be reducing coverage with this, or are we confident that the existing tests were already covering the functionality enough?

We can be confident. The pools tests weren't testing much other than just adding coverage to the pools package. From coverage perspective, everything added is used by archive tests.

cpuguy83 commented:
Error looks legit.

dmcgowan committed:
Remove use of bufio in cli import tests

The use of bufio for writing without flushing can lead to an incomplete
writing of the tar and subsequent unexpected EOF when importing.

Signed-off-by: Derek McGowan <derek@mcg.dev>

thaJeztah commented:
I updated the docker/cli PR I had in draft to test this branch, and seeing a failure that looks related;

- https://github.com/docker/cli/pull/5708

```
time="2024-12-25T10:56:00Z" level=error msg="Can't add file /var/folders/95/0ydz4d79163427j3k5crp3fh0000gn/T/cp-test-465366053/file1 to tar: open /var/folders/95/0ydz4d79163427j3k5crp3fh0000gn/T/cp-test-465366053/file1: no such file or directory"
time="2024-12-25T10:56:00Z" level=error msg="Can't close tar writer: archive/tar: missed writing 8 bytes"
--- FAIL: TestRunCopyFromContainerToFilesystem (0.01s)
    cp_test.go:85: assertion failed: error is not nil: unexpected EOF
Error: something went wrong
Error: exit status 33
Error: exit status 130
time="2024-12-25T10:56:01Z" level=error msg="Error waiting for container: removal failed"
time="2024-12-25T10:56:01Z" level=error msg="error waiting for container: no such container: non-existent-container-id"
FAIL
```

thaJeztah commented:
It also looks like docker/cli still has a direct dependency on `pkg/pools`, but on the PR that introduced it, it was suggested that that code should probably be moved to `pkg/archive`, but looks like that never happened; https://github.com/docker/cli/pull/5708#issuecomment-2561836783

dmcgowan commented:
@thaJeztah Opened a fix for the test, definitely looks like a bad test https://github.com/docker/cli/pull/5715

thaJeztah commented:
Merged that fix, and tested this PR in cli, buildx and buildkit with these draft PRs, and looks like CI is happy on all of them 🎉 ;

- https://github.com/docker/cli/pull/5708
- https://github.com/docker/buildx/pull/2868
- https://github.com/moby/buildkit/pull/5602

thaJeztah reviewed (approved):
LGTM
